habitat-hitl/habitat_hitl/core/ui_elements.py

```diff
@@ -0,0 +1,430 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from enum import IntEnum
+from typing import TYPE_CHECKING, Dict, Final, List, Optional
+
+from habitat_hitl.core.user_mask import Mask, Users
+
+if TYPE_CHECKING:
+    from habitat_hitl.core.client_message_manager import ClientMessageManager
+    from habitat_hitl.core.remote_client_state import RemoteClientState
+
+Color = Optional[List[float]]
+
+AUTO: Final[str] = ""
+
+
+@dataclass
+class UIElement:
+    """
+    Base class for all networked UI elements.
+    """
+
+    uid: str
+
+
+@dataclass
+class UICanvas(UIElement):
+    """
+    Canvas properties
+    """
+
+    padding: int
+    backgroundColor: Color
+
+
+@dataclass
+class UILabel(UIElement):
+    """
+    Text label.
+    """
+
+    text: str
+    horizontalAlignment: int
+    fontSize: int
+    bold: bool
+    color: Color
+
+
+@dataclass
+class UIToggle(UIElement):
+    """
+    Toggle button with two labels.
+    """
+
+    enabled: bool
+    toggled: bool
+    textFalse: str
+    textTrue: str
+    color: Color
+    tooltip: str
+
+
+@dataclass
+class UIListItem(UIElement):
+    """
+    List item with two labels.
+    """
+
+    textLeft: str
+    textRight: str
+    fontSize: int
+    color: Color
+
+
+@dataclass
+class UIButton(UIElement):
+    """
+    Clickable button with a text label.
+    """
+
+    enabled: bool
+    text: str
+    color: Color
+
+
+@dataclass
+class UISeparator(UIElement):
+    """
+    Horizontal or vertical line that helps separate content sections.
+    """
+
+
+@dataclass
+class UISpacer(UIElement):
+    """
+    Horizontal or vertical empty space.
+    """
+
+    size: float
+
+
+class HorizontalAlignment(IntEnum):
+    LEFT = 0
+    CENTER = 1
+    RIGHT = 2
+
+
+class VerticalAlignment(IntEnum):
+    TOP = 0
+    CENTER = 1
+    BOTTOM = 2
+
+
+@dataclass
+class UICanvasUpdate:
+    """
+    Set of UI updates for a specific canvas.
+    """
+
+    clear: bool
+    elements: Optional[List[UIElementUpdate]]
+
+
+@dataclass
+class UIElementUpdate:
+    """
+    UI element to be updated or created.
+    """
+
+    canvasProperties: Optional[UICanvas]
+    label: Optional[UILabel]
+    toggle: Optional[UIToggle]
+    button: Optional[UIButton]
+    listItem: Optional[UIListItem]
+    separator: Optional[UISeparator]
+    spacer: Optional[UISpacer]
```

Comment:
This is ugly but does the job.

Ideally I'd like to find a better option for serializing/deserializing an union without too much boilerplate or serializer features that depend on a specific library.

A type string could be used, but it would arguably result in more error-prone boilerplate.